### PR TITLE
Fashion MNIST using improve unified interface

### PR DESCRIPTION
**Pull request by rajeeja**
Demo example for new community models. 
This is a first draft that uses IMPROVE for initialize parameters and candle for checkpointing.

### CODE REVIEW

File: test/fashion-mnist/infer.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+   # Need to get testloader from Part 1.
+   transform = transforms.Compose([
+      transforms.ToTensor(),
+      transforms.Normalize((0.5,), (0.5,))
+   ])
+   
+   # Get the data directory, batch size and other hyperparameters from params 
+   ##IMPROVE 
+   batch_size = params["batch_size"]
+   learning_rate = params["learning_rate"]
```
Comment (wilke) [inline test/fashion-mnist/infer.py:44]: Why are these parameters for inference?

File: test/fashion-mnist/infer.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+   momentum = params["momentum"]
+   dataset_dir = params["data_dir"]
+   
+   # NOTE: using false now for data loading
+   testset = torchvision.datasets.FashionMNIST(root=dataset_dir, train=False, download=False, transform=transform)
+   testloader = torch.utils.data.DataLoader(testset, batch_size=batch_size, shuffle=True)
```
Comment (wilke) [inline test/fashion-mnist/infer.py:50]: Loading test data for inference? I would expect to load the model weights.

File: test/fashion-mnist/infer.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+   # Check if GPU is available, else use CPU
+   device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+   
+   # Create a neural network model
+   model = Net().to(device)
+
+   # Define optimizer
+   optimizer = optim.SGD(model.parameters(), lr=learning_rate, momentum=momentum)
```
Comment (wilke) [inline test/fashion-mnist/infer.py:59]: Same here

File: test/fashion-mnist/infer.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+         correct += (predicted == labels).sum().item()
+
+   print(f'Accuracy on test set: {(100 * correct / total):.2f}%')
+
+
+
+
+## IMPROVE
+# Note some of these are similar to previous section and may be adjusted as per model requirements
+model_infer_params = [
```
Comment (wilke) [inline test/fashion-mnist/infer.py:85]: Which parameter is for loading model weights?

File: test/fashion-mnist/train.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+        x = x.view(x.size(0), -1)
+        x = torch.relu(self.fc1(x))
+        x = torch.relu(self.fc2(x))
+        x = self.fc3(x)
+        return x
+
+## IMPROVE
+def run(params):
+##
+   # # Define transformations for data preprocessing
```
Comment (wilke) [inline test/fashion-mnist/train.py:35]: If this is for data preprocessing, please move it to the preprocessing script.

File: test/fashion-mnist/train.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+         running_loss += loss.item()
+         if i % 200 == 199:
+               print(f'Epoch: {epoch + 1}, Batch: {i + 1}, Loss: {running_loss / 200}')
+               running_loss = 0.0
+      # set loss and ckpt
+      loss = running_loss / len(trainloader)
+      ckpt.ckpt_epoch(epoch, loss)
+   print('Training finished.')
+
```
Comment (wilke) [inline test/fashion-mnist/train.py:98]: I don't see where you export the final weights.

**Review by wilke — CHANGES_REQUESTED** [review]

File: test/fashion-mnist/infer.py (hunk 1 above)
Comment (adpartin) [reply test/fashion-mnist/infer.py:44]: @rajeeja I agree with @wilke, it's weird that these parameters are defined in inference script. Generally, inference is done without knowledge of training settings (e.g., train batch size, learning rate optimizer, etc.). Is there a reason why these are defined here?

**Comment by adpartin** [comment]
@rajeeja @wilke The code structure doesn't follow the IMPROVE structure, so it's hard to go through the code and fix everything. https://jdacs4c-improve.github.io/docs/content/unified_interface.html

**Comment by rajeeja** [comment]
> @rajeeja @wilke The code structure doesn't follow the IMPROVE structure, so it's hard to go through the code and fix everything. https://jdacs4c-improve.github.io/docs/content/unified_interface.html

That link is not helpful. The data is different. This model doesn't use genetics or drug data, so, the structure won't be exactly the same. The .py files are very simple and easy to understand. Also the notebook is very standard fashion-mnist.

File: test/fashion-mnist/infer.py (hunk 2 above)
Comment (rajeeja) [reply test/fashion-mnist/infer.py:50]: See line 72, how are you going to get images to test? 
`outputs = model(images)`

File: test/fashion-mnist/infer.py (hunk 1 above)
Comment (rajeeja) [reply test/fashion-mnist/infer.py:44]: @adpartin @wilke 
1. we need to load the test data in batches -> Hence - batch_size (see line 50)
2. I'm getting the model from the ckpt method, we can get it via other means also. The ckpt object needs optimizer to instantiate -> Hence - optimizer (see line 64)

File: test/fashion-mnist/infer.py (hunk 3 above)
Comment (rajeeja) [reply test/fashion-mnist/infer.py:59]: same as above :) we can get the model weights in some other fashion, do you know how else to get the model weights to perform inferece? -  where optimizer or learning rate is not needed, it is a minor thing and can be ignored, IMO. The overall logic is to get the model weights and infer on what was done in the training step.

File: test/fashion-mnist/infer.py (hunk 4 above)
Comment (rajeeja) [reply test/fashion-mnist/infer.py:85]: I'm loading it from the ckpt files, so no specific parameter. 

this can be done by using a specific directory to save the model weights and load from there, something along the lines of `test_ml_data_dir `

File: test/fashion-mnist/train.py (hunk 5 above)
Comment (rajeeja) [reply test/fashion-mnist/train.py:35]: trains on 60k images, but there are 10k images we need to infer on, those we have to get here. This is load not the entire preprocessing.

File: test/fashion-mnist/train.py (hunk 6 above)
Comment (rajeeja) [reply test/fashion-mnist/train.py:98]: ckpt, does that for you, if you run it once, you will see it in the directory.
see `ckpt_epoch` line 94 and documentation here: https://candle-lib.readthedocs.io/en/latest/api_ckpt_pytorch_utils/_autosummary/candle.ckpt_pytorch_utils.CandleCkptPyTorch.html

File: test/fashion-mnist/infer.py (hunk 1 above)
Comment (wilke) [reply test/fashion-mnist/infer.py:44]: Sorry, but it is unclear why you must load the data for inference in batches. Is this in any way faster than a simple
```
for v in File
  label=infer(v)
```

Comment (wilke) [reply test/fashion-mnist/infer.py:44]: There should be a better way of loading the model. We want the model from the input directory. No optimizer is needed. If this is a problem, I suggest writing a load_model_weights function as a wrapper.

File: test/fashion-mnist/infer.py (hunk 2 above)
Comment (wilke) [reply test/fashion-mnist/infer.py:50]: I don't understand, or is this a problem of naming conventions? We are doing label prediction in this script no testing. Do you have a specific use case in mind?

File: test/fashion-mnist/infer.py (hunk 1 above)
Comment (wilke) [reply test/fashion-mnist/infer.py:44]: @rajeeja @adpartin If we use this as an example, we have to make it clean. This is a great hack but not a sustainable solution. Please come up with a better solution or hide it in a function call. These are constants in this case.

File: test/fashion-mnist/infer.py (hunk 2 above)
Comment (wilke) [reply test/fashion-mnist/infer.py:50]: @rajeeja Any thoughts?

File: test/fashion-mnist/infer.py (hunk 3 above)
Comment (wilke) [reply test/fashion-mnist/infer.py:59]: I agree, get the model weights and infer on any input data in the input data directory. The model weights should be located there as well. The outputs of training are model weights and learning metrics.

File: test/fashion-mnist/train.py (hunk 6 above)
Comment (wilke) [reply test/fashion-mnist/train.py:98]: The final weights should be in the top-level output directory.
